# Coursera Capstone Project - Parts 1 and 3

## Let's see Part 1 First

# Part 1 - Converting Wikipedia html table into a DataFrame

In [1]:
from bs4 import BeautifulSoup as bsoup
from urllib.request import urlopen as uReq
import requests
import lxml
import pandas as pd
from pandas import DataFrame
import numpy as np

In [2]:
my_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [3]:
r=requests.get(my_url)

## Parsing the web html file with BeautifulSoup package

In [4]:
page=bsoup(r.text,"html.parser")
page


<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"3f996570-256b-4f45-92a3-87ba6c847372","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":960187814,"wgRevisionId":960187814,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toronto","Ontario-rel

## 'The htlm 'page' indicates that there is a 'table class' - let's find that table

In [5]:
rtable=page.table
rtable

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td>

In [6]:
results=rtable.find_all('tr')
nrows=len(results)
nrows

181

## Such table (named 'results' contains 181 rows. The first tow is the table header.
## There are 181 rows of data

In [7]:
results[0:5]

[<tr>
 <th>Postal Code
 </th>
 <th>Borough
 </th>
 <th>Neighborhood
 </th></tr>, <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>, <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>]

### The Header of the DataFrame is the first row of data [index 0]

In [8]:
header=results[0].text.split()
header

['Postal', 'Code', 'Borough', 'Neighborhood']

### Let's check some rows in order to prepare to build the loop that will extract all cells into a DataFrame. For example, let's examine row 33

In [9]:
results[33].text

'\nM6E\n\nYork\n\nCaledonia-Fairbanks\n'

In [10]:
results[33].text.split('\n')

['', 'M6E', '', 'York', '', 'Caledonia-Fairbanks', '']

In [11]:
results[33].text.split('\n')

['', 'M6E', '', 'York', '', 'Caledonia-Fairbanks', '']

In [14]:
Postcode=results[33].text.split('\n')[1]
Postcode

'M6E'

In [15]:
Borough=results[33].text.split('\n')[2]
Borough

''

In [16]:
Neighborhood=results[33].text.split('\n')[3]
Neighborhood

'York'

## Iteration loop to extract all cells into a dataframe df

In [17]:
records =[]
n=1
while n < nrows :
    Postcode=results[n].text.split('\n')[1]
    Borough=results[n].text.split('\n')[2]
    Neighborhood=results[n].text.split('\n')[3]
    records.append((Postcode, Borough,Neighborhood))
    n=n+1

df=pd.DataFrame(records, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df.head(5)

PostalCode Borough     Neighbourhood
0        M1A              Not assigned
1        M2A              Not assigned
2        M3A                North York
3        M4A                North York
4        M5A          Downtown Toronto

In [18]:
df.shape

(180, 3)

## Check the tail of dataframe to assure all data is included as in Wikipedia table

In [19]:
df.tail()

PostalCode Borough Neighbourhood
175        M5Z          Not assigned
176        M6Z          Not assigned
177        M7Z          Not assigned
178        M8Z             Etobicoke
179        M9Z          Not assigned

### Remove all rows with Borough = 'Not assigned' string

In [20]:
df[df['Borough']=='Not assigned'].count()


PostalCode       0
Borough          0
Neighbourhood    0
dtype: int64

In [21]:
df1=df[~df.Borough.str.contains("Not assigned")]
df1=df1.reset_index(drop=True)

In [22]:
df1.head(5)

PostalCode Borough     Neighbourhood
0        M1A              Not assigned
1        M2A              Not assigned
2        M3A                North York
3        M4A                North York
4        M5A          Downtown Toronto

In [23]:
df1.tail()

PostalCode Borough Neighbourhood
175        M5Z          Not assigned
176        M6Z          Not assigned
177        M7Z          Not assigned
178        M8Z             Etobicoke
179        M9Z          Not assigned

In [24]:
df1.shape

(180, 3)

### Now, let's replace 'Not assigned' neighborhoods with the name of the Borough

In [25]:
df1.loc[df1['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df1['Borough']

In [26]:
df1.head()

PostalCode Borough     Neighbourhood
0        M1A                          
1        M2A                          
2        M3A                North York
3        M4A                North York
4        M5A          Downtown Toronto

In [27]:
df1.shape

(180, 3)

### As per instructions, we will look for repeated postalcodes and consolidate into one postalcode with the name of the neighbourhoods merged

In [28]:
postalcodes = df1['PostalCode'].nunique()
boroughs = df1['Borough'].nunique()
neighbourhoods= df1['Neighbourhood'].nunique()
print('Unique Postalcodes : ' + str(postalcodes))
print('Unique Boroughs  : '+ str(boroughs))
print('Unique Neighbourhoods  :' + str(neighbourhoods))

Unique Postalcodes : 180
Unique Boroughs  : 1
Unique Neighbourhoods  :11


In [29]:
#df1.groupby(['PostalCode','Borough','Neighbourhood']).size().reset_index(name='Count').head()

### Consolidate the dataframe to each unique PostalCodes and aggregated Neighbourhoods

In [30]:
df1.head()

PostalCode Borough     Neighbourhood
0        M1A                          
1        M2A                          
2        M3A                North York
3        M4A                North York
4        M5A          Downtown Toronto

In [31]:
df2=df1
df2.head()

PostalCode Borough     Neighbourhood
0        M1A                          
1        M2A                          
2        M3A                North York
3        M4A                North York
4        M5A          Downtown Toronto

### Loop Algorithm to extract the consolidate dataframe df2 with unique postalcodes

In [ ]:
n=0

while n < nrows2 :
    post1=df2.iloc[n,0]
    #post1
    m=n+1
    post2=df2.iloc[m,0]
    #post2
    neigh1=df2.iloc[n,2]
    neigh2=df2.iloc[m,2]
    if post1==post2:
        df2.Neighbourhood[n,2] = neigh1=neigh1+','+neigh2
        #df2 = df2[df2.Neighbourhood != 'neigh2']
        df2=df2.drop(df2.index[m])
        nrows2=nrows2-1
        df2 = df2.reset_index(drop=True)
    else:
        n=n+1


#df2 = df2.reset_index(drop=True)
# print(post1, post2, 'Nigh1 is...'+ neigh1,',,,,,,,','Nigh2 is...' + neigh2, n,m)
df2.index

In [32]:
df2.head()

PostalCode Borough     Neighbourhood
0        M1A                          
1        M2A                          
2        M3A                North York
3        M4A                North York
4        M5A          Downtown Toronto

### DataFrame consolidated to unique PostalCodes and sorted by PostalCode

In [33]:
df2.head(15)

PostalCode Borough     Neighbourhood
0         M1A                          
1         M2A                          
2         M3A                North York
3         M4A                North York
4         M5A          Downtown Toronto
5         M6A                North York
6         M7A          Downtown Toronto
7         M8A                          
8         M9A                 Etobicoke
9         M1B               Scarborough
10        M2B                          
11        M3B                North York
12        M4B                 East York
13        M5B          Downtown Toronto
14        M6B                North York

In [34]:
df2.shape

(180, 3)

## End - PART1

# PART 2: Now, let's get the Latitude and Longitude coordenates for every PostalCode and add those columns to the dataframe

In [35]:
import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

In [36]:
url='http://cocl.us/Geospatial_data'
df_pcodes=pd.read_csv(url)
df_pcodes.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Need to rename the column "Postal Code" to "PostalCode" in order to do a proper merger

In [38]:
df_pcodes.columns = ['PostalCode', 'Latitude', 'Longitude']
df_pcodes.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

### Sort our db2 dataframe by PostalCode, then we merge it with the df_pcodes created from the cvs file, based on common column of PostalCodes

In [39]:
df2s=df2.sort_values('PostalCode')
df2s.head()

PostalCode Borough Neighbourhood
0         M1A                      
9         M1B           Scarborough
18        M1C           Scarborough
27        M1E           Scarborough
36        M1G           Scarborough

### dataframe 'neighborhoods' contains the latitude and longitude needed to start mapping

In [42]:
neighborhoods=pd.merge(df2s,df_pcodes, how='right', on = 'PostalCode')
neighborhoods.head(15)

PostalCode Borough Neighbourhood   Latitude  Longitude
0         M1B           Scarborough  43.806686 -79.194353
1         M1C           Scarborough  43.784535 -79.160497
2         M1E           Scarborough  43.763573 -79.188711
3         M1G           Scarborough  43.770992 -79.216917
4         M1H           Scarborough  43.773136 -79.239476
5         M1J           Scarborough  43.744734 -79.239476
6         M1K           Scarborough  43.727929 -79.262029
7         M1L           Scarborough  43.711112 -79.284577
8         M1M           Scarborough  43.716316 -79.239476
9         M1N           Scarborough  43.692657 -79.264848
10        M1P           Scarborough  43.757410 -79.273304
11        M1R           Scarborough  43.750072 -79.295849
12        M1S           Scarborough  43.794200 -79.262029
13        M1T           Scarborough  43.781638 -79.304302
14        M1V           Scarborough  43.815252 -79.284577

## End of PART 2

# PART 3

## Explore and cluster the neighborhoods in Toronto

In [40]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2020.4.5.2         |   py36h9f0ad1d_0         152 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    ca-certificates-2020.4.5.2 |       hecda079_0         147 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    ------------------------------------------------------------
                       

### There are 38 Borough with word Toronto

In [45]:
toronto_data= neighborhoods[neighborhoods['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)
toronto_data.head()

Empty DataFrame
Columns: [PostalCode, Borough, Neighbourhood, Latitude, Longitude]
Index: []

In [46]:
toronto_data.shape

(0, 5)

### From google : the latitud and longitud for Toronto, as follows: 43.6532° N, 79.3832° W

In [49]:
latitude = 43.6532
longitude= -79.3832

### Create a map of Toronto with Boroughs that cointain the word Toronto

In [51]:
# create map of TORONTO using latitude and longitude values above:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [73]:
CLIENT_ID = 'DISWDLDS53WKKYNKQFPHEWMGR2WPZUHBYCEQIL335I1Z4WCF' # your Foursquare ID
CLIENT_SECRET = 'KEE5XHMTGPSEY5JHRC3ME1BK5XUTQTQR5FXKKJ0H1EBMTERP' # your Foursquare Secret

### Create a function to repeat the same process to all the neighborhoods in Manhattan Lab example

In [75]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radiu

In [77]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Now write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues

In [ ]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

In [ ]:
print(toronto_venues.shape)
toronto_venues.head()

In [ ]:
toronto_venues.groupby('Neighborhood').count()

In [ ]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

In [ ]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols=list(toronto_onehot.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehot=toronto_onehot[['Neighborhood']+cols]

# rename Neighborhood for Neighbourhood so that future merge works
toronto_onehot.rename(columns = {'Neighborhood': 'Neighbourhood'}, inplace = True)
toronto_onehot.head()

In [ ]:
toronto_onehot.shape

In [ ]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

In [ ]:
toronto_grouped.shape

### Let's print each neighborhood along with the top 5 most common venues

In [ ]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

### Let's put that into a pandas dataframe

In [ ]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

## Cluster Neighbourhoods

In [ ]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

In [ ]:
toronto_merged = toronto_data

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

In [ ]:
## Final resulting clusters visualization - Toronto Neighborhoods